arcs/src/components/spatial_entity.rs

```diff
@@ -0,0 +1,145 @@
+        let bb = self.bounds;
+        TypedRect::<f32, f64>::new(
+            // TypedRects have their origin at the bottom left corner (this is undocumented!)
+            TypedPoint2D::new(bb.bottom_left().x as f32, bb.bottom_left().y as f32),
+            TypedSize2D::new(bb.width() as f32, bb.height() as f32))
+    }
+}
+
+impl SpatialEntity {
+    pub fn new(bounds: BoundingBox, entity: Entity) -> SpatialEntity {
+        SpatialEntity {
+            bounds,
+            entity
+        }
+    }
+}
+
+/// A global [`Resource`] for looking up which [`Entity`]s inhabit
+/// a given spatial point or region
+#[derive(Debug)]
+pub struct Space {
+    quadtree: SpatialTree,
+    ids: HashMap<Index, ItemId>
+}
+
+impl Default for Space {
+    fn default() -> Self {
+        Space {
+            quadtree: Self::default_tree(),
+            ids: HashMap::new()
+        }
+    }
+}
+
+impl Space {
+    // FIXME: Hard-code is bad-bad
+    const WORLD_RADIUS: f64 = 1_000_000.0;
+
+    // FIXME: We need to supply this in *DrawingUnits*
+    // as 1[meter] f.e. becomes meaningless when zoomed far in/out
+    const QUERY_POINT_RADIUS: f64 = 1.0;
+
+    fn default_tree() -> SpatialTree{
+        // Initialize quadtree
+        let size = BoundingBox::new(
+            Vector::new(-Self::WORLD_RADIUS, -Self::WORLD_RADIUS),
+            Vector::new(Self::WORLD_RADIUS, Self::WORLD_RADIUS)
+            ).aabb();
+        let quadtree: SpatialTree = QuadTree::new(
+            size,
+            true,
+            4,
+            16,
+            8,
+            4
+        );
+
+        quadtree
+    }
+
+    pub fn insert(&mut self, spatial: SpatialEntity) {
+        let entity_id = spatial.entity.id();
+        if let Some(id) = self.quadtree.insert(spatial) {
+            self.ids.insert(entity_id, id);
+        }
+    }
+
+    pub fn modify(&mut self, spatial: SpatialEntity) {
+        let entity_id = spatial.entity.id();
+        if self.ids.contains_key(&entity_id) {
+            let item_id = self.ids[&entity_id];
+
+            // remove old item
+            self.quadtree.remove(item_id);
+            self.ids.remove(&entity_id);
+
+            // Add modified
+            self.insert(spatial);
+        }
+    }
+
+    pub fn remove_by_id(&mut self, id: Index) {
+        if self.ids.contains_key(&id) {
+            let item_id = self.ids[&id];
+
+            // remove old item
+            self.quadtree.remove(item_id);
+            self.ids.remove(&id);
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        self.ids.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.ids.is_empty()
+    }
+
+    pub fn query_point(&self, point: Vector) -> Option<Vec<Entity>> {
+        let cursor_circle = Arc::from_centre_radius(point, Self::QUERY_POINT_RADIUS, 0.0, 2.0 * std::f64::consts::PI);
+        self.query_region(cursor_circle.bounding_box())
+    }
+
+    pub fn query_region(&self, region: BoundingBox) -> Option<Vec<Entity>> {
```

Comment:
Does the underlying quad tree provide an API for searching lazily (i.e. by returning an iterator) instead of greedily collecting into a `Vec`? 

From my experience, a lot of the time you're only interested in the first result matching a predicate so it's a waste of time/memory to populate a vector with all objects within the region. Returning an `impl Iterator<Item = SpatialEntity>` would also feel a bit more idiomatic.

Reply:
the [query](https://docs.rs/aabb-quadtree/0.2.0/aabb_quadtree/struct.QuadTree.html#method.query) function is hardcoded to return a `small-vec`, maybe it would be possible using the `custom_query` method, I can look into that.

Also +1 on `impl Iterator<Item = SpatialEntity>`

Reply:
changed return type to `impl Iterator<Item = SpatialEntity>`